### ISSUE TITLE
Single Entry 404 when View is embedded on page with /search/ slug

### ISSUE THREAD

**Issue by zackkatz** [gravityview/GravityView#809]
Changing the slug to anything else fixes the problem.

See https://secure.helpscout.net/conversation/286720465/0/

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/995529792029955/995647703014340)

**Comment by soulseekah** [comment]
https://secure.helpscout.net/conversation/105661465/2914/

**Comment by soulseekah** [comment]
Also related, but this is a "view" slug instead of "search" https://secure.helpscout.net/conversation/498316192/12727

**Comment by soulseekah** [comment]
Reproduced:

![image](https://user-images.githubusercontent.com/685880/69901062-52e0c480-138d-11ea-98b0-a6876102793b.png)

**Comment by soulseekah** [comment]
"search" is a reserved source, I'm not entirely sure we should be messing with it by reordering ourselves:

```
wp rewrite list --match='/search/entry/123'
+--------------------------+--------------------------------------------------+--------+
| match                    | query                                            | source |
+--------------------------+--------------------------------------------------+--------+
| search/(.+)/?$           | index.php?s=$matches[1]                          | search |
| (.?.+?)/entry(/(.*))?/?$ | index.php?pagename=$matches[1]&entry=$matches[3] | page   |
| (.?.+?)(?:/([0-9]+))?/?$ | index.php?pagename=$matches[1]&page=$matches[2]  | page   |
+--------------------------+--------------------------------------------------+--------+
```

I don't think anyone should ever be naming their pages like that. While it matches fine like this:

```
wp rewrite list --match='/search/'         
+--------------------------+-------------------------------------------------+--------+
| match                    | query                                           | source |
+--------------------------+-------------------------------------------------+--------+
| (.?.+?)(?:/([0-9]+))?/?$ | index.php?pagename=$matches[1]&page=$matches[2] | page   |
+--------------------------+-------------------------------------------------+--------+
```

If anything else is added after `/search/` WordPress matches the search page almost always:

```
wp rewrite list --match='/search/sdf'
+--------------------------+-------------------------------------------------+--------+
| match                    | query                                           | source |
+--------------------------+-------------------------------------------------+--------+
| search/(.+)/?$           | index.php?s=$matches[1]                         | search |
| (.?.+?)(?:/([0-9]+))?/?$ | index.php?pagename=$matches[1]&page=$matches[2] | page   |
+--------------------------+-------------------------------------------------+--------+
```

We could try to put `| (.?.+?)/entry(/(.*))?/?$ | index.php?pagename=$matches[1]&entry=$matches[3] | page   |` higher up as a priority and it sort of will work but might have side effects. One of which would be that searching the site for "something something /entry/1" will not return a search results page as it should.

Will anyone ever search for something that includes "hello /entry/1"? Or are the chances of someone creating a page with a "search" slug more probable? Should we be discouraging naming pages with reserved keywords?

Or should we hook into the 404 page when the search contains "entry/1" and output a notice to admins that accessing single entries on a "search" slug is a bad idea and here's why, linking to a doc?

**Comment by zackkatz** [comment]
We should show a warning when saving the view when the slug matches a reserved source.

**Comment by soulseekah** [comment]
> We should show a warning when saving the view when the slug matches a reserved source.

But we don't know whether the View we are saving is embedded in a page with a /search/
I'm not even talking about saving a View with a reserved slug which is a different but very related issue. Haven't thought about it.

On a page or a post or a CPT (this includes a View type) that has a /search/ slug. If administrator show a warning?

**Comment by zackkatz** [comment]
@soulseekah Yes, any time we know things will break due to configuration, we should show show a warning.

**Comment by soulseekah** [comment]
- [ ] Reserve "search" (`WP_Rewrite::$search_base`), "view", "entry" slugs (latter two according to https://docs.gravityview.co/article/57-customizing-urls)
- [ ] direct View access
- [ ] shortcodes
- [ ] oEmbeds
- [ ] tests

**Comment by soulseekah** [comment]
The documentation is at https://secure.helpscout.net/docs/566dfabbc6979143615564bf/article/5de618682c7d3a7e9ae4a707/ and should be published so we can get a public URL for the warning message. @zackkatz

**Comment by soulseekah** [comment]
Manual testing looks good. Unit tests look good, will finalize and push tomorrow.

I propose a merge after 2.5 is released, unless you want to include it in 2.5 after a bit of further testing :)

I am, however, very sure that there will be other cases of this. I suspect we'll see this pop up in date archives, category slugs, etc. permalinks are a pain here. I've used up too much time on this one issue for now (testing, fiddling, trying to break things), we can return to it once something similar is encountered in the wild.

**Comment by soulseekah** [comment]
This is done.

For @zackkatz:

- [ ] check and modify language strings as needed
- [ ] check and publish documentation
- [ ] set documentation URL in the warning message
- [ ] PR and merge

whew.. this was a tough one
